Solutions/Corvus.Storage.Azure.Cosmos/Corvus/Storage/Azure/Cosmos/CosmosContainerFactory.cs

```diff
@@ -86,65 +76,60 @@ protected override string GetCacheKeyForParentConfiguration(CosmosContainerConfi
             return this.GetCacheKeyForConfiguration(nonContainerSpecificConfiguration);
         }
 
-        private async ValueTask<CosmosClient> CreateCosmosClientAsync(
+        /// <inheritdoc/>
+        protected override async ValueTask InvalidateForConfigurationAsync(
+            CosmosContainerConfiguration contextConfiguration,
+            IAzureTokenCredentialSource? tokenCredentialSource,
+            CosmosClientOptions? connectionOptions,
+            CancellationToken cancellationToken)
+        {
+            if (tokenCredentialSource is not null)
+            {
+                await tokenCredentialSource.GetReplacementForFailedTokenCredentialAsync(cancellationToken)
```

Comment:
Am I right in thinking that the intention is that Invalidate pays to recreate rather than just invalidating? Is this definitely the best name?

Reply:
Having slept on this I am now wondering if we need to add an alternate way to invalidate `ClientIdentityConfiguration` in `Corvus.Identity`: the need to hang on to the specific instances of `IAzureTokenCredentialSource` that were returned has added a great deal of complication. I think if `Corvus.Identity` were to let you just pass in another equivalent `ClientIdentityConfiguration` and say "the credentials you're handing out for this no longer seem to be any good".

That would offer two benefits here:
1. we would no longer force the creation of new credentials here
2. I could take back out all the complexity I just added to be able to hold onto the objects returned by `Corvus.Identity`

I'm not sure 1 is that big a deal in practice: the scenario in which we expect this to be called is "I'm trying to use this storage service and have just found out that I can't, so can we please see if any keys got changed and try again?" So we're going to be creating new credentials anyway. It would just change where it happens.

But I agree it's still a bit unsatisfactory, and 2 is a bigger deal I think.

Reply:
OK, I've created https://github.com/corvus-dotnet/Corvus.Identity/pull/226 to try this out, and I've updated this PR to use it - I think this improves this situation